File: epi_judge_java/epi/PrimitiveDivide.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

public class PrimitiveDivide {
  @EpiTest(testDataFile = "primitive_divide.tsv")
  public static int divide(int x, int y) {
    int res = 0;
    while (x >= y) {
      int shift = 0;
      long yPower = y;
      while (x >= (yPower << shift)) {
        shift++;
      }
      shift--;
      res += 1 << shift;
      x -= y << shift;
    }

    return res;
  }

  public static void main(String[] args) {
    int res = divide(2097428739, 186);
    System.exit(
        GenericTest
            .runFromAnnotations(args, "PrimitiveDivide.java",
                new Object() {
                }.getClass().getEnclosingClass())
            .ordinal());
  }
}

File: epi_judge_python/primitive_divide.py
from test_framework import generic_test


def divide(x: int, y: int) -> int:
    res = 0
    while x >= y:
        shift = 1
        while x >= (y << shift):
            shift += 1
        shift -= 1
        res += (1 << shift)
        x -= (y << shift)
    return res


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('primitive_divide.py',
                                       'primitive_divide.tsv', divide))
